llama-index-core/llama_index/core/llms/function_calling.py

```diff
@@ -0,0 +1,165 @@
+from typing import (
+    Any,
+    List,
+    Optional,
+    Union,
+    TYPE_CHECKING,
+)
+import asyncio
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+)
+from llama_index.core.llms.llm import LLM
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+)
+from llama_index.core.llms.llm import ToolSelection
+
+if TYPE_CHECKING:
+    from llama_index.core.chat_engine.types import AgentChatResponse
+    from llama_index.core.tools.types import BaseTool
+
+
+class FunctionCallingLLM(LLM):
+    """
+    Function calling LLMs are LLMs that support function calling.
+    They support an expanded range of capabilities.
+
+    """
+
+    def chat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Predict and call the tool."""
+        raise NotImplementedError("predict_tool is not supported by default.")
+
+    async def achat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Predict and call the tool."""
+        raise NotImplementedError("predict_tool is not supported by default.")
+
+    def get_tool_calls_from_response(
+        self,
+        response: "AgentChatResponse",
+        error_on_no_tool_call: bool = True,
+        **kwargs: Any,
+    ) -> List[ToolSelection]:
+        """Predict and call the tool."""
+        raise NotImplementedError(
+            "get_tool_calls_from_response is not supported by default."
+        )
+
+    def predict_and_call(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> "AgentChatResponse":
+        """Predict and call the tool."""
+        from llama_index.core.chat_engine.types import AgentChatResponse
+        from llama_index.core.tools.calling import (
+            call_tool_with_selection,
+        )
+
+        if not self.metadata.is_function_calling_model:
```

Comment:
ooc: in our LLMMetadata class this attribute is whether the LLM has a function calling API similar to OpenAI's function call api. Does that description need to be changed to something more general. In other words, does Mistral use a similar function call api to OpenAI?

Reply:
Actually I am bit confused as to why we need this from `metadata.is_function_calling_model` in this case. 

I mean if this is `FunctionCallingLLM`, then shouldn't `self.metadata.is_function_calling_model` always be `True`?

Reply:
I think there's a difference between having function calling built into the API (mistral, openai) vs using some prompting to orchestrate tool calling (the default react loop)